File: backend/controllers/methodPayController.js
const PhuongThucThanhToan = require("../models/PhuongThucThanhToan");
const fs = require("fs");
const path = require("path");

// Lấy danh sách đang hoạt động
exports.getActivePayments = async (req, res) => {
  try {
    const payments = await PhuongThucThanhToan.findAll({
      where: { trang_thai: "active" },
      order: [["id", "ASC"]],
    });
    res.status(200).json(payments);
  } catch (error) {
    res.status(500).json({ message: "Lỗi server" });
  }
};

// Lấy tất cả
exports.getAllPaymentsAdmin = async (req, res) => {
  try {
    const payments = await PhuongThucThanhToan.findAll({
      order: [["id", "DESC"]],
    });
    res.status(200).json(payments);
  } catch (error) {
    res.status(500).json({ message: "Lỗi server" });
  }
};

// Thêm mới
exports.createPayment = async (req, res) => {
  try {
    const { ten_phuong_thuc, ma, loai, trang_thai } = req.body || {};

    let logo_url = null;
    if (req.file) {
      logo_url = `/uploads/${req.file.filename}`;
    }

    const newPayment = await PhuongThucThanhToan.create({
      ten_phuong_thuc,
      ma,
      loai,
      logo_url,
      trang_thai: trang_thai || "active",
    });

    res.status(201).json({ message: "Thêm thành công", data: newPayment });
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: "Lỗi khi thêm phương thức" });
  }
};

// Sửa
exports.updatePayment = async (req, res) => {
  try {
    const { id } = req.params;
    const { ten_phuong_thuc, ma, loai, trang_thai } = req.body || {};

    const payment = await PhuongThucThanhToan.findByPk(id);
    if (!payment) return res.status(404).json({ message: "Không tìm thấy" });

    payment.ten_phuong_thuc = ten_phuong_thuc || payment.ten_phuong_thuc;
    payment.ma = ma || payment.ma;
    payment.loai = loai || payment.loai;
    if (trang_thai) payment.trang_thai = trang_thai;

    if (req.file) {
      if (payment.logo_url) {
        const oldPath = path.join(__dirname, "../", payment.logo_url);
        if (fs.existsSync(oldPath)) fs.unlinkSync(oldPath);
      }
      payment.logo_url = `/uploads/${req.file.filename}`;
    }

    await payment.save();
    res.status(200).json({ message: "Cập nhật thành công" });
  } catch (error) {
    res.status(500).json({ message: "Lỗi cập nhật" });
  }
};

// Bật/Tắt
exports.togglePaymentStatus = async (req, res) => {
  try {
    const payment = await PhuongThucThanhToan.findByPk(req.params.id);
    payment.trang_thai =
      payment.trang_thai === "active" ? "inactive" : "active";
    await payment.save();
    res.status(200).json({ message: "Đã đổi trạng thái" });
  } catch (error) {
    res.status(500).json({ message: "Lỗi server" });
  }
};

// Xóa
exports.deletePayment = async (req, res) => {
  try {
    const payment = await PhuongThucThanhToan.findByPk(req.params.id);
    if (payment.logo_url) {
      const filePath = path.join(__dirname, "../", payment.logo_url);
      if (fs.existsSync(filePath)) fs.unlinkSync(filePath);
    }
    await payment.destroy();
    res.status(200).json({ message: "Xóa thành công" });
  } catch (error) {
    res.status(500).json({ message: "Lỗi xóa" });
  }
};

File: backend-csharp/Controllers/MethodPayController.cs
using System;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using WebFashion.Api.Models;

namespace WebFashion.Api.Controllers
{
    [Route("api/payments")]
    [ApiController]
    public class MethodPayController : ControllerBase
    {
        private readonly AppDbContext _context;

        public MethodPayController(AppDbContext context)
        {
            _context = context;
        }

        private bool IsAdmin()
        {
            var userRole = User.FindFirst("vai_tro")?.Value;
            return userRole == "admin";
        }

        // 1. GET: api/payments
        [HttpGet]
        [AllowAnonymous]
        public async Task<IActionResult> GetActivePayments()
        {
            try
            {
                var list = await _context.PhuongThucThanhToans
                    .Where(p => p.TrangThai == "active")
                    .OrderBy(p => p.Id)
                    .ToListAsync();

                return Ok(list);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi lấy phương thức thanh toán active: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi server" });
            }
        }

        // 2. GET: api/payments/admin-pay
        [HttpGet("admin-pay")]
        [Authorize]
        public async Task<IActionResult> GetAllPaymentsAdmin()
        {
            try
            {
                if (!IsAdmin())
                {
                    return StatusCode(403, new { message = "Quyền truy cập bị từ chối!" });
                }

                var list = await _context.PhuongThucThanhToans
                    .OrderByDescending(p => p.Id)
                    .ToListAsync();

                return Ok(list);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi lấy phương thức thanh toán admin: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi server" });
            }
        }

        // 3. POST: api/payments
        [HttpPost]
        [Authorize]
        [Consumes("multipart/form-data")]
        public async Task<IActionResult> CreatePayment(
            [FromForm] string ten_phuong_thuc,
            [FromForm] string ma,
            [FromForm] string loai,
            [FromForm] string? trang_thai,
            IFormFile? logo_url)
        {
            try
            {
                if (!IsAdmin())
                {
                    return StatusCode(403, new { message = "Quyền truy cập bị từ chối!" });
                }

                string? logoPath = null;
                if (logo_url != null && logo_url.Length > 0)
                {
                    var uploadsDir = Path.Combine(Directory.GetCurrentDirectory(), "uploads");
                    if (!Directory.Exists(uploadsDir))
                    {
                        Directory.CreateDirectory(uploadsDir);
                    }

                    var uniqueFileName = $"{Guid.NewGuid()}_{Path.GetFileName(logo_url.FileName)}";
                    var filePath = Path.Combine(uploadsDir, uniqueFileName);

                    using (var stream = new FileStream(filePath, FileMode.Create))
                    {
                        await logo_url.CopyToAsync(stream);
                    }

                    logoPath = $"/uploads/{uniqueFileName}";
                }

                var newPayment = new PhuongThucThanhToan
                {
                    TenPhuongThuc = ten_phuong_thuc,
                    Ma = ma,
                    Loai = loai,
                    LogoUrl = logoPath,
                    TrangThai = trang_thai ?? "active",
                    PhiThanhToan = 0m
                };

                _context.PhuongThucThanhToans.Add(newPayment);
                await _context.SaveChangesAsync();

                return StatusCode(201, new { message = "Thêm thành công", data = newPayment });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi thêm phương thức thanh toán: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi khi thêm phương thức" });
            }
        }

        // 4. PUT: api/payments/{id}
        [HttpPut("{id}")]
        [Authorize]
        [Consumes("multipart/form-data")]
        public async Task<IActionResult> UpdatePayment(
            long id,
            [FromForm] string? ten_phuong_thuc,
            [FromForm] string? ma,
            [FromForm] string? loai,
            [FromForm] string? trang_thai,
            IFormFile? logo_url)
        {
            try
            {
                if (!IsAdmin())
                {
                    return StatusCode(403, new { message = "Quyền truy cập bị từ chối!" });
                }

                var payment = await _context.PhuongThucThanhToans.FindAsync(id);
                if (payment == null)
                {
                    return NotFound(new { message = "Không tìm thấy" });
                }

                if (!string.IsNullOrEmpty(ten_phuong_thuc)) payment.TenPhuongThuc = ten_phuong_thuc;
                if (!string.IsNullOrEmpty(ma)) payment.Ma = ma;
                if (!string.IsNullOrEmpty(loai)) payment.Loai = loai;
                if (!string.IsNullOrEmpty(trang_thai)) payment.TrangThai = trang_thai;

                if (logo_url != null && logo_url.Length > 0)
                {
                    var uploadsDir = Path.Combine(Directory.GetCurrentDirectory(), "uploads");
                    if (!Directory.Exists(uploadsDir))
                    {
                        Directory.CreateDirectory(uploadsDir);
                    }

                    var uniqueFileName = $"{Guid.NewGuid()}_{Path.GetFileName(logo_url.FileName)}";
                    var filePath = Path.Combine(uploadsDir, uniqueFileName);

                    using (var stream = new FileStream(filePath, FileMode.Create))
                    {
                        await logo_url.CopyToAsync(stream);
                    }

                    // Delete old logo
                    if (!string.IsNullOrEmpty(payment.LogoUrl))
                    {
                        var oldFilePath = Path.Combine(Directory.GetCurrentDirectory(), payment.LogoUrl.TrimStart('/'));
                        if (System.IO.File.Exists(oldFilePath))
                        {
                            try
                            {
                                System.IO.File.Delete(oldFilePath);
                            }
                            catch { }
                        }
                    }

                    payment.LogoUrl = $"/uploads/{uniqueFileName}";
                }

                await _context.SaveChangesAsync();

                return Ok(new { message = "Cập nhật thành công", data = payment });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi cập nhật phương thức thanh toán: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi cập nhật" });
            }
        }

        // 5. PUT: api/payments/{id}/toggle
        [HttpPut("{id}/toggle")]
        [Authorize]
        public async Task<IActionResult> TogglePaymentStatus(long id)
        {
            try
            {
                if (!IsAdmin())
                {
                    return StatusCode(403, new { message = "Quyền truy cập bị từ chối!" });
                }

                var payment = await _context.PhuongThucThanhToans.FindAsync(id);
                if (payment == null)
                {
                    return NotFound(new { message = "Không tìm thấy" });
                }

                payment.TrangThai = payment.TrangThai == "active" ? "inactive" : "active";
                await _context.SaveChangesAsync();

                return Ok(new { message = "Đã đổi trạng thái", trang_thai = payment.TrangThai });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi toggle payment status: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi server" });
            }
        }

        // 6. DELETE: api/payments/{id}
        [HttpDelete("{id}")]
        [Authorize]
        public async Task<IActionResult> DeletePayment(long id)
        {
            try
            {
                if (!IsAdmin())
                {
                    return StatusCode(403, new { message = "Quyền truy cập bị từ chối!" });
                }

                var payment = await _context.PhuongThucThanhToans.FindAsync(id);
                if (payment == null)
                {
                    return NotFound(new { message = "Không tìm thấy" });
                }

                // Delete logo file
                if (!string.IsNullOrEmpty(payment.LogoUrl))
                {
                    var filePath = Path.Combine(Directory.GetCurrentDirectory(), payment.LogoUrl.TrimStart('/'));
                    if (System.IO.File.Exists(filePath))
                    {
                        try
                        {
                            System.IO.File.Delete(filePath);
                        }
                        catch { }
                    }
                }

                _context.PhuongThucThanhToans.Remove(payment);
                await _context.SaveChangesAsync();

                return Ok(new { message = "Xóa thành công" });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi xóa phương thức thanh toán: {ex.Message}");
                return StatusCode(500, new { message = "Lỗi xóa" });
            }
        }
    }
}
